### PR TITLE
advertise and discover added to memory transport

### PR DESCRIPTION
**Pull request by zippy**
## PR summary

- adds discovery to in-memory server
- current solution is to have only the first discovery ever do anything because after initial discovery further nodes should be found by gossip
- adds pattern of passing network-id to transport for it to advertise

## Review checklist 
- [ ] The story has unit or integration tests
- [ ] No new bugs, and any tech-debt is identified and justified
- [ ] There is enough API documentation (how to use)
- [ ] There is enough code documentation (how the code works)

### CODE REVIEW

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs [hunk 1]
```diff
@@ -39,20 +44,55 @@ pub type GhostTransportMemoryEndpointContextParent = GhostContextEndpoint<
...
+    fn advertise(&mut self) -> DiscoveryResult<()> {
+        let uri = self
+            .maybe_my_address
+            .clone()
+            .ok_or_else(|| DiscoveryError::new_other("must bind before discovering"))?;
+        let mut verse = memory_server::get_memory_verse();
+        let _ = verse
+            .get_net(&self.network)
+            .ok_or_else(|| DiscoveryError::new_other("net not found"))?
+            .advertise(uri, self.machine_id.clone());
```
Comment (neonphog) [inline crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: `?`

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs [hunk 2]
```diff
@@ -63,6 +103,52 @@ impl GhostTransportMemory {
...
+                                    Some(remote_server) => {
+                                        let _result = remote_server.request_connect(&my_addr);
+                                        self.connections.insert(found_uri.clone());
+                                        trace!("Discovered {}, we are: {}", &found_uri, &my_addr);
+                                        let _ = self.endpoint_self.publish(
+                                            Lib3hSpan::fixme(),
+                                            RequestToParent::IncomingConnection {
+                                                uri: found_uri.clone(),
+                                            },
+                                        );
```
Comment (neonphog) [inline crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: `?`

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs [hunk 3]
```diff
@@ -182,7 +271,7 @@ impl
                         verse.bind(&self.network)
                     };
                     self.maybe_my_address = Some(bound_url.clone());
-
+                    let _ = self.advertise();
```
Comment (struktured) [inline crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: Confusing here- should we check for errors? Or at least give the binding a descriptive name?

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs [hunk 4]
```diff
@@ -63,6 +103,52 @@ impl GhostTransportMemory {
             ),
             connections: HashSet::new(),
             maybe_my_address: None,
+            last_discover: start,
+            discover_interval_ms: interval as u128,
+        }
+    }
+
+    fn try_discover(&mut self) {
+        if let Some(t) = self.last_discover {
```
Comment (struktured) [inline crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: This code block is a bit overly nested but probably not worth cleaning up at this point.

File: crates/lib3h/src/transport/memory_mock/memory_server.rs [hunk 5]
```diff
@@ -29,15 +29,23 @@ pub enum MemoryEvent {
...
 impl MemoryNet {
     pub fn new() -> Self {
         MemoryNet {
             server_map: HashMap::new(),
             url_count: 0,
+            advertised_machines: HashSet::new(),
         }
     }
+    pub fn advertise(&mut self, uri: Url, machine_id: Address) {
+        let _ = self.advertised_machines.insert((uri, machine_id));
```
Comment (struktured) [inline crates/lib3h/src/transport/memory_mock/memory_server.rs]: Probably don't ignore this.

File: crates/lib3h/src/engine/ghost_engine.rs [hunk 6]
```diff
@@ -216,12 +194,10 @@ impl<'engine> GhostEngine<'engine> {
...
         for bs in nodes {
             // can't use handle_bootstrap() because it assumes a message to respond to
-            let cmd = GatewayRequestToChild::Transport(
-                transport::protocol::RequestToChild::SendMessage {
-                    uri: bs,
-                    payload: Opaque::new(),
-                },
-            );
+            let cmd = GatewayRequestToChild::Bootstrap(BootstrapData {
+                space_address: "bogus_space".into(),
```
Comment (struktured) [inline crates/lib3h/src/engine/ghost_engine.rs]: Why is the space bogus?

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs [hunk 7]
```diff
@@ -39,20 +44,55 @@ pub type GhostTransportMemoryEndpointContextParent = GhostContextEndpoint<
...
 pub struct GhostTransportMemory {
+    machine_id: Address,
     network: String,
     endpoint_parent: Option<GhostTransportMemoryEndpoint>,
     endpoint_self: Detach<GhostTransportMemoryEndpointContext>,
     /// My peer uri on the network layer (not None after a bind)
     maybe_my_address: Option<Url>,
     /// Addresses of connections to remotes
     connections: HashSet<Url>,
+    last_discover: Option<Instant>,
```
Comment (struktured) [inline crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: You might be able to get rid of the option and just set this to a sufficiently max or min value, but perhaps what you have here is clearer.

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs [hunk 8]
```diff
@@ -63,6 +103,52 @@ impl GhostTransportMemory {
             ),
             connections: HashSet::new(),
             maybe_my_address: None,
+            last_discover: start,
+            discover_interval_ms: interval as u128,
```
Comment (struktured) [inline crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: `u128` is cool but feels like overkill.

File: crates/lib3h/src/transport/websocket/actor.rs [hunk 9]
```diff
@@ -7,26 +7,47 @@ use crate::transport::{
...
     bound_url: Option<Url>,
     pending: Vec<Message>,
 }
 
+impl Discovery for GhostTransportWebsocket {
+    fn advertise(&mut self) -> DiscoveryResult<()> {
+        Ok(())
+    }
+    fn discover(&mut self) -> DiscoveryResult<Vec<Url>> {
+        //        let nodes: Vec<Url> = self.server_map.keys().map(|x| x.clone()).collect();
```
Comment (struktured) [inline crates/lib3h/src/transport/websocket/actor.rs]: Remove comment.

File: crates/lib3h/src/transport/websocket/actor.rs [hunk 10]
```diff
@@ -7,26 +7,47 @@ use crate::transport::{
...
+use lib3h_protocol::{data_types::Opaque, Address};
 use lib3h_tracing::Lib3hSpan;
 use url::Url;
 
 pub type Message =
     GhostMessage<RequestToChild, RequestToParent, RequestToChildResponse, TransportError>;
 
 pub struct GhostTransportWebsocket {
+    #[allow(dead_code)]
+    machine_id: Address,
```
Comment (struktured) [inline crates/lib3h/src/transport/websocket/actor.rs]: Temporarily dead, I assume?

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs [hunk 11]
```diff
@@ -39,20 +44,55 @@ pub type GhostTransportMemoryEndpointContextParent = GhostContextEndpoint<
...
+        Ok(())
+    }
 }
 
 impl GhostTransportMemory {
     #[allow(dead_code)]
-    pub fn new(network_name: &str) -> Self {
+    pub fn new(machine_id: Address, network_name: &str) -> Self {
         let (endpoint_parent, endpoint_self) = create_ghost_channel();
+        let interval = 30000;
```
Comment (struktured) [inline crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: Eventually should b a config parameter of engine or at the very least make a well named constant for this?

File: crates/lib3h/src/engine/ghost_engine.rs (hunk 6 above)
Comment (zippy) [reply crates/lib3h/src/engine/ghost_engine.rs]: added comment.  this should come from the config, but that space address (equiv to DNA address) isn't there yet.

File: crates/lib3h/src/transport/websocket/actor.rs (hunk 10 above)
Comment (zippy) [reply crates/lib3h/src/transport/websocket/actor.rs]: yep

File: crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs (hunk 11 above)
Comment (zippy) [reply crates/lib3h/src/transport/memory_mock/ghost_transport_memory.rs]: added

**Pull request merged**